src/lib/cpuid/cpuid_factory.cpp

```diff
@@ -6,47 +6,54 @@
 #include "cpuid/cpuid_native_config.h"
 #include "cpuid/cpuid_native.h"
 
-namespace rjcp::cpuid {
-
-template<>
-CpuIdFactory<CpuIdDefaultConfig>::CpuIdFactory() noexcept = default;
+namespace rjcp::cpuid
+{
 
-template<>
-CpuIdFactory<CpuIdDefaultConfig>::CpuIdFactory(CpuIdDefaultConfig& /* config */) noexcept
+namespace
 {
-    // There is no configuration.
-}
+template <typename CreationCallback>
+class CpuIdFactoryWithCallback : public ICpuIdFactory
+{
+public:
+    CpuIdFactoryWithCallback(const CreationCallback &create_callback) : create_callback_{create_callback} {};
 
-template<>
-auto CpuIdFactory<CpuIdDefaultConfig>::create(int cpunum) noexcept -> std::unique_ptr<ICpuId>
+    auto create(int cpunum) noexcept -> std::unique_ptr<ICpuId> override
+    {
+        return create_callback_(cpunum);
+    }
+
+private:
+    CreationCallback create_callback_;
+};
+
+template <typename CreationCallback>
+auto MakeCpuIdFactory(const CreationCallback &callback) -> std::unique_ptr<ICpuIdFactory>
 {
-    return std::make_unique<CpuIdDefault>(cpunum);
+    return std::make_unique<CpuIdFactoryWithCallback<CreationCallback>>(callback);
 }
+} // namespace
 
-template<>
-CpuIdFactory<CpuIdNativeConfig>::CpuIdFactory() noexcept = default;
+// TODO: Consider extracting the CreateCpuIdFactory specializations into seperate translation units to decouple the concrete types.
```

Comment:
To do this, wouldn't the [unnamed namespace](https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces) need to be removed as per [internal linkage](https://en.cppreference.com/w/cpp/language/storage_duration#Linkage) to be visible at link time?

> The name can be referred to from all scopes in the current translation unit.